java-frontend/src/main/java/org/sonar/java/SonarComponents.java

```diff
@@ -494,67 +493,75 @@ public InputComponent project() {
+    if (!isAutoScan()) {
+      // In autoscan, test + main code are analyzed in the same batch, and we do not make the distinction between
+      // test and main libraries, everything is inside "sonar.java.libraries", it is expected to let the test property empty.
+      javaTestClasspath.logSuspiciousEmptyLibraries();
     }
+    logUndefinedTypes(LOGGED_MAX_NUMBER_UNDEFINED_TYPES);
+
+    // clear the set so only new undefined types will be logged
+    problemsToFilePaths.clear();
   }
 
   private void logUndefinedTypes(int maxLines) {
     logParserMessages(
-      undefinedTypes.stream()
-        .filter(m -> m.type() == JProblem.Type.UNDEFINED_TYPE),
+      problemsToFilePaths.entrySet().stream()
+        .filter(entry -> entry.getKey().type() == JProblem.Type.UNDEFINED_TYPE),
       maxLines,
       "Unresolved imports/types have been detected during analysis. Enable DEBUG mode to see them.",
       "Unresolved imports/types:"
     );
     logParserMessages(
-      undefinedTypes.stream()
-        .filter(m -> m.type() == JProblem.Type.PREVIEW_FEATURE_USED),
+      problemsToFilePaths.entrySet().stream()
+        .filter(entry -> entry.getKey().type() == JProblem.Type.PREVIEW_FEATURE_USED),
       maxLines,
       "Use of preview features have been detected during analysis. Enable DEBUG mode to see them.",
       "Use of preview features:"
     );
   }
 
-  private static void logParserMessages(Stream<JProblem> messages, int maxLines, String warningMessage, String debugMessage) {
-    final List<String> messagesList = messages
-      .map(Object::toString)
-      .sorted()
+  private static void logParserMessages(Stream<Map.Entry<JProblem, List<String>>> messages, int maxProblems, String warningMessage, String debugMessage) {
+    final String problemDelimiter = System.lineSeparator() + "- ";
+    final List<List<String>> messagesList = messages
+      .sorted(Comparator.comparing(entry -> entry.getKey().toString()))
+      .limit(maxProblems + 1L)
+      .map(entry -> {
+        List<String> paths = entry.getValue();
+        List<String> problemAndPaths = new ArrayList<>(paths.size() + 1);
+        problemAndPaths.add(problemDelimiter + entry.getKey().toString());
+        paths.forEach(path -> problemAndPaths.add("  * " + path));
+        return problemAndPaths;
+      })
       .collect(Collectors.toList());
-    int messagesListSize = messagesList.size();
-    if (messagesListSize == 0) {
-      return;
-    }
-    final boolean moreThanMax = messagesListSize > maxLines;
 
-    if (moreThanMax) {
-      debugMessage += " (Limited to " + maxLines + ")";
+    if (messagesList.isEmpty()) {
+      return;
     }
 
-    final String delimiter = System.lineSeparator() + "- ";
-    final String prefix = debugMessage + delimiter;
-    final String suffix = moreThanMax ? (delimiter + "...") : "";
-
     LOG.warn(warningMessage);
     if (LOG.isDebugEnabled()) {
+      final boolean moreThanMax = messagesList.size() > maxProblems;
```

Comment:
not sure about the `final` modifier here. Either we use it everywhere, or not. In practice, we almost never use it in our projects, so I would drop it.